ERTP API Alpha Release update and clean up. (#249)
Agoric/documentation: main/ertp/api/purse.md

# Purse
`Purses` hold digital assets. They are created to hold assets of a single `brand`,
and their `brand` cannot be changed. For example, you might create a `purse` to
hold Quatloos. That `purse` can only hold Quatloos; it can't hold Moola or theater
tickets or anything else. And you can't change that `purse` so that it now holds
Moola instead; you need to create a `purse` that holds only `moola`.

Digital assets in `purses` and `payments` can be any of:
- Currency-like, such as our imaginary Quatloos currency.
- Goods-like digital assets, such as magic weapons for use in a game or theater tickets.
- Other kinds of rights, such as the right to participate in a particular contract.

While each `purse` can only hold assets of one `brand`, any number of `purses` can be
created that hold that `brand`. So you could have, say, three Quatloos `purses`, your
friend Alice could have eight Quatloos `purses`, and so on. 

You change a `purse`'s balance by calling either `deposit()` (to add assets)
or `withdraw()` (to remove assets) on it. A `purse` can be empty, meaning it

> has a balance value of 0 (which depending on the kind of asset, could also be
> that just aren't any, say, theater tickets in the `purse`).

Reviewer: ```suggestion
has a balance value of 0 (which, for a ticket-holding purse, means that
there just aren't any theater tickets in the `purse`).
```

Unlike `payments`, `purses` are not meant to be sent to others. 
To transfer digital assets, you should withdraw a `payment` from a `purse` and
send the `payment` to another party.

## issuer.makeEmptyPurse()
- Returns: `{Purse}`

While not a method called on a `purse`, clearly it's important to know how
to create a new `purse`. Call `makeEmptyPurse()` on the `issuer` associated
with the `brand` of assets you want the new `purse` to hold. It returns the
new `purse` with an empty balance. 
```js
const {issuer: quatloosIssuer} = makeIssuerKit('quatloos');
quatloosPurse1 = quatloosIssuer.makeEmptyPurse();
quatloosPurse2 = quatloosIssuer.makeEmptyPurse();
```

## purse.getCurrentAmount()
- Returns: `{Amount}`

Get an `amount` describing the current digital assets balance in the `purse`.
This `amount` is confirmed by the `issuer` for the `brand` the `purse` can hold.
Of course, the returned `amount` `value` might be different the next time you
call `getCurrentAmount()` on the same `purse` if assets have been deposited or

> withdrawn from it in-between calls. . 

Author: Good to know, but can the current doc text stand? I think so, but want to check.  I think this ends up being more a digression on asynchronous programming, which perhaps should be added to the "Diestributed Programming with JavaScript" catchall programming tips doc, but I don't think fits here. Could easily be wrong though; I'm not clear on whether this was meant as educational text for me or if you'd like this included in some form in the text. Thanks.

> withdrawn from it in-between calls. . 

Reviewer: Mostly yes, the current text can stand. I think the fact that `deposit()` uniquely requires a non-promise (and the implications) should be called out explicitly in the entry for `deposit()`.

```js
const { issuer: quatloosIssuer } = makeIssuerKit('quatloos');
const quatloosPurse = quatloosIssuer.makeEmptyPurse();
// quatloos5 is a payment with balance of 5 quatloos
const quatloosPurse.deposit(quatloos5);
// Returns an amount with value = 5 and brand = quatloos
const currentBalance = quatloosPurse.getCurrentAmount();
```

## purse.deposit(payment, optAmount)
- `payment` `{Payment}`
- `optAmount` `{Amount}` - Optional. 
- Returns: `{Amount}`

Deposit all the contents of `payment` into this `purse`, returning an `amount` describing the
`payment`'s digital assets (i.e. the deposit amount). If the optional argument `optAmount` does not equal the balance of
`payment`, or if `payment` is an unresolved promise, it throws an error.

```js
const { issuer: quatloosIssuer, mint: quatloosMint, amountMath: quatloosAmountMath } = 
      makeIssuerKit('quatloos');
const quatloosPurse = quatloosIssuer.makeEmptyPurse();
const payment = quatloosMint.mintPayment(quatloosAmountMath.make(123));
const quatloos123 = quatloosAmountMath.make(123);

// Deposit a payment for 123 Quatloos into the purse. Ensure that this is the amount you expect.

> quatloosPurse.deposit(payment, quatloos123);

Reviewer: ```suggestion
const depositAmount = quatloosPurse.deposit(payment, quatloos123);
```

One has to look at the return value to find out if the deposit succeeded. I dunno if you want to add a line of code checking that the returned value was 123.

const secondPayment = quatloosMint.mintPayment(quatloosAmountMath.make(100));
// Throws error
quatloosPurse.deposit(secondPayment, quatloos123);

```

## purse.makeDepositFacet()

Creates a deposit-only facet on the `purse`. This is an object you can give to other parties
that lets them deposit `payments` to your  `purse` without being able to withdraw assets from or check
the balance of the `purse`. This makes it a safe way to let other people send you `payments`.

You can only deposit a `payment` into a deposit facet that's the same `brand` as the original `purse`
takes.
 
```js
const depositOnlyFacet = purse.makeDepositFacet();

// Give depositOnlyFacet to someone else. Anyone with a deposit facet reference can tell it to receive
// a `payment`, thus depositing the `payment` assets in the `purse` associated with the deposit facet.
depositOnlyFacet.receive(payment);
```
Once you have created a `depositFacet`, there is one method you can call 

> on it. **tyg todo: is it "recieve" or "deposit"?** `deposit.receive(payment)`. 

Reviewer: it is `receive`

The `depositFacet` takes the `payment` and adds it to the balance of the facet's
associated `purse`. The `payment` must be of the same `brand` as what the `purse` holds.

## purse.withdraw(amount)
- `amount` `{Amount}`
- Returns: `{Payment}`

Withdraw the `amount` specified digital assets from this `purse` into a new `payment`.

```js
// Create a purse and give it a balance of 10 Quatloos
const { issuer, mint } = makeIssuerKit('quatloos');
const purse = issuer.makeEmptyPurse();
const payment = mint.mintPayment(amountMath.make(10));
const quatloos10 = amountMath.make(10);
purse.deposit(payment, quatloos10);

> // Withdraw 3 amount from the purse

Reviewer: ```suggestion
// Withdraw an amount of 3 from the purse
```

# Purse
`Purses` hold digital assets. They are created to hold assets of a single `brand`,
and their `brand` cannot be changed. For example, you might create a `purse` to
hold Quatloos. That `purse` can only hold Quatloos; it can't hold Moola or theater
tickets or anything else. And you can't change that `purse` so that it now holds
Moola instead; you need to create a `purse` that holds only Moola.

Digital assets in `purses` and `payments` can be any of:
- Currency-like, such as our imaginary Quatloos currency.
- Goods-like digital assets, such as magic weapons for use in a game or theater tickets.
- Other kinds of rights, such as the right to participate in a particular contract.

While each `purse` can only hold assets of one `brand`, any number of `purses` can be
created that hold that `brand`. So you could have, say, three Quatloos `purses`, your
friend Alice could have eight Quatloos `purses`, and so on. 

You change a `purse`'s balance by calling either `deposit()` (to add assets)
or `withdraw()` (to remove assets) on it. A `purse` can be empty, meaning it
has a balance value of 0 (which, for a ticket-holding `purse`, means that

> there just aren't any theater tickets in the `purse`).

Reviewer 2: I can see that we're trying to accommodate the nonfungible side here too, but can we remove the value of 0? Technically, the value could be []. So although I can see what we're trying to do, I think it is technically incorrect.
```suggestion
or `withdraw()` (to remove assets) on it. A `purse` can be empty. (A purse for a currency
has a balance value of 0, but a ticket-holding purse just doesn't have any tickets).
```

Author: How about adding "fungible" to have "A purse for a fungible currency has a balance value of 0, but..." to make the difference clear?

Reviewer 2: sounds good!

hold Quatloos. That `purse` can only hold Quatloos; it can't hold Moola or theater
tickets or anything else. And you can't change that `purse` so that it now holds
Moola instead; you need to create a `purse` that holds only Moola.

Digital assets in `purses` and `payments` can be any of:
- Currency-like, such as our imaginary Quatloos currency.
- Goods-like digital assets, such as magic weapons for use in a game or theater tickets.
- Other kinds of rights, such as the right to participate in a particular contract.

While each `purse` can only hold assets of one `brand`, any number of `purses` can be
created that hold that `brand`. So you could have, say, three Quatloos `purses`, your
friend Alice could have eight Quatloos `purses`, and so on. 

You change a `purse`'s balance by calling either `deposit()` (to add assets)
or `withdraw()` (to remove assets) on it. A `purse` can be empty, which if it holds 
a fungible currency means it has a value of 0. If it holds non-fungible theatre tickets, 
it means it just doesn't have any tickets.

meaning it
has a balance value of 0 (which, for a ticket-holding `purse`, means that

> there just aren't any theater tickets in the `purse`).

Reviewer: redundant

friend Alice could have eight Quatloos `purses`, and so on. 

You change a `purse`'s balance by calling either `deposit()` (to add assets)
or `withdraw()` (to remove assets) on it. A `purse` can be empty, which if it holds 
a fungible currency means it has a value of 0. If it holds non-fungible theatre tickets, 
it means it just doesn't have any tickets.

Unlike `payments`, `purses` are not meant to be sent to others. 
To transfer digital assets, you should withdraw a `payment` from a `purse` and
send the `payment` to another party.

## issuer.makeEmptyPurse()
- Returns: `{Purse}`

While not a method called on a `purse`, clearly it's important to know how
to create a new `purse`. Call `makeEmptyPurse()` on the `issuer` associated
with the `brand` of assets you want the new `purse` to hold. It returns the
new `purse` with an empty balance. 
```js
const {issuer: quatloosIssuer} = makeIssuerKit('quatloos');
quatloosPurse1 = quatloosIssuer.makeEmptyPurse();
quatloosPurse2 = quatloosIssuer.makeEmptyPurse();
```

## purse.getCurrentAmount()
- Returns: `{Amount}`

Get an `amount` describing the current digital assets balance in the `purse`.

> This `amount` is confirmed by the `issuer` for the `brand` the `purse` can hold.

Reviewer 2: ```suggestion
```

Reviewer 2: This doesn't make sense. I think if this came from the types files, we have an issue on agoric-sdk to remove it.

Of course, the returned `amount` `value` might be different the next time you
call `getCurrentAmount()` on the same `purse` if assets have been deposited or
withdrawn from it in-between calls. . 

```js
const { issuer: quatloosIssuer } = makeIssuerKit('quatloos');
const quatloosPurse = quatloosIssuer.makeEmptyPurse();
// quatloos5 is a payment with balance of 5 quatloos
const quatloosPurse.deposit(quatloos5);
// Returns an amount with value = 5 and brand = quatloos
const currentBalance = quatloosPurse.getCurrentAmount();
```

## purse.deposit(payment, optAmount)
- `payment` `{Payment}`
- `optAmount` `{Amount}` - Optional. 
- Returns: `{Amount}`

> All I came up with that is specific to deposit was this:

Reviewer 2: Is this an internal comment that accidentally got in??

Reviewer: yep, that's my note to Tom. Shouldn't be here.

> 1) you have to have a non-promise before you can call deposit(). When you call deposit() you'll get a response back (after waiting only for the round trip) telling you whether it succeeded. All calls that arrive later will see that the value has been transferred into the purse, and the payments are no longer valid.

Reviewer 2: ```suggestion

```
Remove? I think this was an accident

You can only deposit a `payment` into a deposit facet that's the same `brand` as the original `purse`
takes.
 
```js
const depositOnlyFacet = purse.makeDepositFacet();

// Give depositOnlyFacet to someone else. Anyone with a deposit facet reference can tell it to receive
// a payment, thus depositing the payment assets in the purse associated with the deposit facet.
depositOnlyFacet.receive(payment);
```
Once you have created a `depositFacet`, there is one method you can call 
on it. `depositFacet.receive(payment)`. The `depositFacet` takes the `payment` 
and adds it to the balance of the facet's associated `purse`. The `payment` 
must be of the same `brand` as what the `purse` holds.

Note the difference in method names for adding assets between a `purse` and its `depositFacet`.
To add assets to a `purse` directly, you use `purse.deposit()`. To add assets
to a `purse` via its `depositFacet`, you use `depositFacet.receive()`.

## purse.withdraw(amount)
- `amount` `{Amount}`
- Returns: `{Payment}`

Withdraw the `amount` specified digital assets from this `purse` into a new `payment`.

If the call succeeds, it immediately extracts the value into a new `payment`. 

> The caller won't get the new `payment` until a later turn, since the call is (nearly always) remote.
> But as soon as the message is processed, the value is gone from the `purse`.

Reviewer 2: Hmm, where is this coming from? if we are calling it in this way `purse.withdraw`, then we are assuming the purse is local, so the caller gets the result synchronously, in the same turn. If we are thinking the purse is remote, then we need to call it like `E(purse).withdraw` I think it's best to stick to the purse being local, which means we should remove this
```suggestion
```

Author: It's coming from a Keybase conversation with Chris about one of his comments. 
Me: "You made a comment in purse.md " I think the fact that deposit() uniquely requires a non-promise (and the implications) should be called out explicitly in the entry for deposit()."   What are the implications, so I can add them to the deposit() section? Thanks."

Reviewer 2: Ok, I'm wondering if we should be adding so much new material when we're just trying to get this out. This new material has nuances, as I mentioned above. In the rest of the doc, we describe the purse as if it is local. Chris' analysis assumes it is remote. We will need to explain why we are suddenly shifting perspective.

Reviewer: Why would it make sense to present the API as if most interactions are with a local purse? There are certainly aspects of the interaction for which the distinction isn't relevant but nearly all interactions are with remote purses, so giving developers an awareness of the asynchronous behavior where it's pertinent seems important.

Reviewer 2: It's certainly correct that most interactions will be with a remote purse. I'm just pointing out that this document, with this new  inclusion, will be remarkably inconsistent. We need to explain that this particular analysis is for a remote purse, and our instructions previously (in which we mention calling `purse.deposit` or `purse.withdraw`) only work on local purses. If the purse is remote, we must do `E(purse).deposit` and `E(purse).withdraw`. It's the inconsistency that is the problem, not the analysis of remote purses.
